crowbar_framework/app/models/service_object.rb

```diff
@@ -1826,4 +1826,35 @@ def skip_unready_nodes(bc, inst, new_elements, old_elements)
     logger.debug("skip_unready_nodes: exit for #{bc}:#{inst}")
     [cleaned_elements, pre_cached_nodes]
   end
+
+  # return true if the new attributes are different from the old ones
+  def node_changed_attributes?(node, old_role, new_role)
+    if old_role.default_attributes[@bc_name] != new_role.default_attributes[@bc_name]
+      logger.debug("#{@bc_name} skip_batch_for_node?: not skipping #{node} (attribute change)")
+      return true
+    end
+
+    false
+  end
+
+  # return true if the node has changed roles
+  def node_changed_roles?(node, old_role, new_role)
+    role_constraints.keys.each do |role_name|
```

Comment:
As said previously, you shouldn't use `role_constraints` but directly the content of `old_role` and `new_role`.

Something like:

```ruby
roles_in_old = old_role.elements.keys.select { |r| old_role.elements[r].include?(node) }.sort
roles_in_new = new_role.elements.keys.select { |r| new_role.elements[r].include?(node) }.sort
roles_in_old != roles_in_new
```

(completely untested)

Reply:
I have tested it and works.